### ISSUE TITLE
Add filter to get the host or :authority

### ISSUE THREAD

**Issue by Restioson** [seanmonstar/warp#432]
It appears that when using TLS with Warp, the HTTP Host header is stripped. Therefore, `warp::header<String>("host")` on a TLS server will fail.

Ironically, the Haskell framework with the same name had an identical bug: https://github.com/yesodweb/wai/issues/478

**Comment by jxs** [comment]
I don't think this is a bug, when warp is serving using TLS, clients offer Http2 using ALPN, hyper then upgrades the connection to Http2, and per the Http2 [spec](https://http2.github.io/http2-spec/):
>Clients that generate HTTP/2 requests directly SHOULD use the :authority pseudo-header field instead of the Host header field.

accessing the host via [`Uri::authority()` ](https://docs.rs/hyper/0.13.2/hyper/struct.Uri.html#method.authority) doesn't work for you?
you can get it constructing an `Uri` with [`warp::path::full()`](https://docs.rs/warp/0.2.1/warp/filters/path/fn.full.html)

**Comment by Restioson** [comment]
Thanks, I will try both when I get home.

**Comment by Restioson** [comment]
I first tried extracting `:authority` and `authority` using the `warp::header` method, to be greeted by identical behaviour of requiring `host`. For reference, here are my Firefox raw HTTPS request headers:

```
Host: 127.0.0.1:8080
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko/20100101 Firefox/71.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Upgrade-Insecure-Requests: 1
Cache-Control: max-age=0
TE: Trailers
```

Other info:
```
Request URL:https://127.0.0.1:8080/invite/4XwS5ZEoVHU
Request Method:GET
Remote Address:127.0.0.1:8080
Status Code:
400
Version:HTTP/2.0
```

However, `warp::path::full()` does work! Thanks!

**Comment by Restioson** [comment]
Oops, nevermind. It looks like `warp::path::full` just returns the path rather than the entire URI (makes sense, given its name).

**Comment by jxs** [comment]
ah! you are right! `warp::path::full` doesn't return the `Authority` 
hum, this could solved by making [`FullPath`](https://github.com/seanmonstar/warp/blob/master/src/filters/path.rs#L420) be 
```Rust
struct FulllPath(String)
```  
instead 
and [`warp::reply::full`](https://github.com/seanmonstar/warp/blob/master/src/filters/path.rs#L415) be
```Rust

**Comment by seanmonstar** [comment]
We could add a `host()` filter which checks the URI authority and if not present, checks for the host header (RFC 7230 specifies that order).

Probably return some newtype `Host`, wrapping a `http::uri::Authority`?

**Comment by jxs** [comment]
ah yeah, totally forgot that changing `warp::reply::full` behavior would be a breaking change, `host()` filter makes more sense :+1:

**Comment by freddyb** [comment]
This is considered fixed due to the issues mentioned above this comment, right?

**Comment by theduke** [comment]
This issue can be closed.

( https://docs.rs/warp/0.3.1/warp/filters/host/index.html )

**Comment by jxs** [comment]
solved with #676
